### DISCUSSION TITLE
Using System.Text.Json deserialization issues

### DISCUSSION THREAD

**Discussion by talfirevic** [JasperFx/marten#2483] [Q&A]
I've set up `System.Text.Json` as a default serialized by implementing `ISerializer` and I'm having issues with a basic use case - store and retrieve an entity.

this is the relevant part of the CustomSerializer

```
 public string ToJson(object? document)
    {
        return JsonSerializer.Serialize(document);
    }

    public T FromJson<T>(Stream stream)
    {
        return JsonSerializer.Deserialize<T>(stream);
    }
```

Storing an object invokes `ToJson`, and it works as intended.

Querying that same entity gets the item from the database, but when `FromJson<T>` is invoked, deserialization breaks because the string starts with `0x01` for some odd reason.

I'm missing something very obvious, aren't I?

P.S. when the object is being stored, the serializer does not produce a JSON string with  `0x01` at the beginning, it only happens when I retrieve the object from the database.

**Comment by Hawxy** [comment]
Is there a reason you're implementing this yourself? There's already [built in support for STJ](https://martendb.io/configuration/json.html#serialization-with-system-text-json) and it handles this particular problem.

**Comment by talfirevic** [comment]
@Hawxy , I'm aware this is already implemented. I plan on using a different serializer (likely MessagePack) when I get the time to deal with the whole serialization topic in the app.

By "this particular problem" were you referring to the `0x01` thing or STJ serialization in general?

**Comment by Hawxy** [comment]
> I plan on using a different serializer (likely MessagePack)

I'm not sure how that would work. Marten uses jsonb selectors and thus serializing to encoded MessagePack would break a lot of functionality.

> By "this particular problem" were you referring to the 0x01 thing or STJ serialization in general?

The `0x01` problem. Marten's STJ serializer has a custom stream implementation that skips the control character:
https://github.com/JasperFx/marten/blob/64598221e07050a8d1d82ef855011994e094260a/src/Marten/Services/SystemTextJsonSerializer.cs

**Comment by jeremydmiller** [comment]
@talfirevic MessagePack would never work for the document storage. That would completely break the Linq support.

**Comment by jeremydmiller** [comment]
I'm switching this to a discussion as there's not anything to "fix" in Marten itself. 

@talfirevic My advice is to just use the built in STJ serialization strategy here that was built to be able to support Linq querying.
